File: leetcode/go/src/leetcode/113_path_sum_2.go
package leetcode

func innerPathSum(root *TreeNode, sum int, path *[]int, res *[][]int) {
	if root == nil {
		return
	}
	*path = append(*path, root.Val)
	if root.Left == nil && root.Right == nil && sum == root.Val {
		*res = append(*res, *path)
	}
	if root.Left != nil {
		newpath := make([]int, len(*path))
		copy(newpath, *path)
		innerPathSum(root.Left, sum-root.Val, &newpath, res)
	}
	if root.Right != nil {
		newpath := make([]int, len(*path))
		copy(newpath, *path)
		innerPathSum(root.Right, sum-root.Val, &newpath, res)
	}
}

func pathSum(root *TreeNode, sum int) [][]int {
	var path []int
	var res [][]int
	innerPathSum(root, sum, &path, &res)
	return res
}

File: leetcode/cpp/113_path_sum2.cpp
class Solution {
 public:
  // time: O(n), space: O(n^2)
  vector<vector<int>> pathSum(TreeNode *root, int sum) {
    vector<vector<int>> res;
    vector<int> init;
    innerPath(root, sum, init, res);
    return res;
  }

  void innerPath(TreeNode *root, int sum, vector<int> &saved,
                 vector<vector<int>> &all) {
    if (root != nullptr && root->left == nullptr && root->right == nullptr &&
        root->val == sum) {
      saved.push_back(root->val);
      all.push_back(saved);
    }
    if (root != nullptr && (root->left != nullptr || root->right != nullptr)) {
      saved.push_back(root->val);
      if (nullptr != root->left) {
        auto vec = saved;
        innerPath(root->left, sum - root->val, vec, all);
      }
      if (nullptr != root->right) {
        auto vec = saved;
        innerPath(root->right, sum - root->val, vec, all);
      }
    }
  }
};
